src/Generics/SOP/Universe.hs

```diff
@@ -152,3 +153,34 @@ class HasDatatypeInfo a where
   datatypeInfo         :: proxy a -> DatatypeInfo (Code a)
   default datatypeInfo :: (GDatatypeInfo a, GCode a ~ Code a) => proxy a -> DatatypeInfo (Code a)
   datatypeInfo = gdatatypeInfo
+
+-- | Constraint that captures that a datatype is a product type,
+-- i.e., a type with a single constructor.
+--
+-- It also gives access to the code for the arguments of that
+-- constructor.
+--
+type IsProductType (a :: *) (xs :: [*]) =
+  (Generic a, Code a ~ '[ xs ])
+
+-- | Constraint that captures that a datatype is an enumeration type,
+-- i.e., none of the constructors have any arguments.
+--
+type IsEnumType (a :: *) =
+  (Generic a, All ((~) '[]) (Code a))
```

Comment:
Yes, I agree that the `Nat` variant would be consistent. If there's ever a desire to have this, I wouldn't be opposed to change it.